signal.go

```diff
@@ -38,17 +45,104 @@ func (sig ShutdownSignal) String() string {
 }
 
 func newSignalReceivers() signalReceivers {
-	return signalReceivers{notify: signal.Notify}
+	return signalReceivers{
+		notify:  signal.Notify,
+		signals: make(chan os.Signal, 1),
+	}
 }
 
 type signalReceivers struct {
-	m      sync.Mutex
-	last   *ShutdownSignal
-	done   []chan os.Signal
+	// this mutex protects writes and reads of this struct to prevent
+	// race conditions in a parallel execution pattern
+	m sync.Mutex
+
+	// our os.Signal channel we relay from
+	signals chan os.Signal
+	// when written to, will instruct the signal relayer to shutdown
+	shutdown chan struct{}
+	// is written to when signal relay has finished shutting down
+	finished chan struct{}
+
+	// this stub allows us to unit test signal relay functionality
 	notify func(c chan<- os.Signal, sig ...os.Signal)
+
+	// last will contain a pointer to the last ShutdownSignal received, or
+	// nil if none, if a new channel is created by Wait or Done, this last
+	// signal will be immediately written to, this allows Wait or Done state
+	// to be read after application stop
+	last *ShutdownSignal
+
+	// contains channels created by Done
+	done []chan os.Signal
+
+	// contains channels created by Wait
+	wait []chan ShutdownSignal
+}
+
+func (recv *signalReceivers) relayer(ctx context.Context) {
+	defer func() {
+		recv.finished <- struct{}{}
+	}()
+
+	select {
+	case <-recv.shutdown:
+		return
+	case <-ctx.Done():
+		return
+	case signal := <-recv.signals:
+		recv.Broadcast(ShutdownSignal{
+			Signal: signal,
+		})
+	}
+}
+
+func (recv *signalReceivers) running() bool {
+	return recv.shutdown != nil && recv.finished != nil
+}
+
+func (recv *signalReceivers) Start(ctx context.Context) {
+	recv.m.Lock()
+	defer recv.m.Unlock()
+
+	// if the receiver has already been started; don't start it again
+	if recv.running() {
+		return
+	}
+
+	recv.last = nil
+	recv.finished = make(chan struct{}, 1)
+	recv.shutdown = make(chan struct{}, 1)
+	recv.notify(recv.signals, os.Interrupt, _sigINT, _sigTERM)
+	go recv.relayer(ctx)
+}
+
+func (recv *signalReceivers) Stop(ctx context.Context) error {
+	recv.m.Lock()
+	defer recv.m.Unlock()
+
+	// if the relayer is not running; return nil error
+	if !recv.running() {
+		return nil
+	}
+
+	recv.shutdown <- struct{}{}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-recv.finished:
+			close(recv.shutdown)
+			close(recv.finished)
+			recv.shutdown = nil
+			recv.finished = nil
+			return nil
+		}
+
+	}
```

Comment:
This doesn't need to be in a infinite loop. 
```suggestion
	select {
	case <-ctx.Done():
		return ctx.Err()
	case <-recv.finished:
		close(recv.shutdown)
		close(recv.finished)
		recv.shutdown = nil
		recv.finished = nil
		return nil
	}
```